Objects/dictobject.c

```diff
@@ -2379,7 +2379,7 @@ dict_merge(PyObject *a, PyObject *b, int override)
         return -1;
     }
     mp = (PyDictObject*)a;
-    if (PyDict_Check(b)) {
+    if (PyDict_CheckExact(b)) {
```

Comment:
This may be too restrictive. Most dict subclasses don't change the order. Maybe check also that `__iter__` is not overridden?

Reply:
You mean, "Maybe check instead", right?  So something like the following.

```c
if (PyDict_Check(b) && Py_TYPE(b)->tp_iter == PyDict_Type.tp_iter) {
```

Reply:
Adding the tp_iter checks seems like the most reasonable way to go. That would preserve the fast conversion speed of defaultdict.